apps/mymedicare_cb/models.py

```diff
@@ -1,76 +1,120 @@
+        AssertionError: If a user is matched but not all identifiers match.
+    """
+    subject = user_info['sub']
+    hicn = user_info['hicn']
+    hicn_hash = hash_hicn(hicn)
+
     try:
-        user = User.objects.get(username=username)
-        if not user.first_name:
-            user.first_name = user_info['given_name']
-        if not user.last_name:
-            user.last_name = user_info['family_name']
-        if not user.email:
-            user.email = user_info['email']
-        user.save()
+        user = User.objects.get(username=subject)
+        assert user.crosswalk.user_id_hash == hicn_hash, "Found user's hicn did not match"
+        return user
     except User.DoesNotExist:
-        # Create a new user. Note that we can set password
-        # to anything, because it won't be checked.
-        user = User(username=username, password='',
-                    first_name=user_info['given_name'],
-                    last_name=user_info['family_name'],
-                    email=user_info['email'])
-        user.set_unusable_password()
-        user.save()
-    UserProfile.objects.get_or_create(
-        user=user, user_type='BEN')
-    group = Group.objects.get(name='BlueButton')
-    user.groups.add(group)
-    # Log in the user
-    user.backend = 'django.contrib.auth.backends.ModelBackend'
-
-    # Determine patient_id
-    fhir_source = get_resourcerouter()
-    crosswalk, _ = Crosswalk.objects.get_or_create(
-        user=user, fhir_source=fhir_source)
-    hicn = user_info.get('hicn', "")
-    crosswalk.set_hicn(hicn)
-    crosswalk.save()
+        pass
+
+    first_name = user_info.get('given_name', "")
+    last_name = user_info.get('family_name', "")
+    email = user_info.get('email', "")
 
     try:
-        backend_data = authenticate_crosswalk(crosswalk)
+        fhir_id, backend_data = match_hicn_hash(hicn_hash)
         # Get first and last name from FHIR if not in OIDC Userinfo response.
-        if user_info['given_name'] == "" or user_info['family_name'] == "":
-            if 'entry' in backend_data:
-                if 'name' in backend_data['entry'][0]['resource']:
-                    names = backend_data['entry'][0]['resource']['name']
-                    first_name = ""
-                    last_name = ""
-                    for n in names:
-                        if n['use'] == 'usual':
-                            last_name = n['family']
-                            first_name = n['given'][0]
-                        if last_name or first_name:
-                            user.first_name = first_name
-                            user.last_name = last_name
-                            user.save()
-    except (UpstreamServerException, exceptions.NotFound):
-        logger.error("Failed to connect Beneficiary "
-                     "to FHIR")
+        if first_name == "" or last_name == "":
+            first_name, last_name = extract_beneficiary_names(backend_data)
```

Comment:
I agree this is weird. It is the original behavior, but I'm going to remove it completely for the moment in order to reduce complexity. It can be added back alter if a business reason arises.